File: lib/core/crypto/field_crypto_key_store.dart
import 'package:flutter/foundation.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_secure_storage/flutter_secure_storage.dart';

import '../../features/auth/domain/user_session.dart';
import '../config/app_config.dart';
import '../database/app_database.dart';
import 'field_crypto.dart';

final fieldCryptoKeyStoreProvider = Provider<FieldCryptoKeyStore>((ref) {
  return FieldCryptoKeyStore(
    ref.watch(appDatabaseProvider),
    const FlutterSecureStorage(),
  );
});

/// AES field-crypto key: Drift `local_app_kv` only (never Firestore).
/// Prefers [AppConfig.fieldCryptoKeyBase64] when set (shared org key).
class FieldCryptoKeyStore {
  FieldCryptoKeyStore(this._db, this._secure);

  static const kvKey = 'field_crypto_key_v1';
  static const _legacySecureKey = 'salik_field_crypto_key_v1';

  final AppDatabase _db;
  final FlutterSecureStorage _secure;

  FieldCrypto? _cached;

  FieldCrypto? get current => _cached;

  Future<FieldCrypto?> ensureKey(UserSession? session) async {
    if (_cached != null) return _cached;

    final fromDefine = AppConfig.fieldCryptoKeyBase64.trim();
    if (fromDefine.isNotEmpty) {
      final existing = await _db.getKv(kvKey);
      if (existing != fromDefine) {
        await _db.setKv(kvKey, fromDefine);
      }
      return _cacheFromBase64(fromDefine);
    }

    final fromDrift = await _db.getKv(kvKey);
    if (fromDrift != null && fromDrift.isNotEmpty) {
      return _cacheFromBase64(fromDrift);
    }

    // One-shot migrate from older FlutterSecureStorage installs.
    try {
      final legacy = await _secure.read(key: _legacySecureKey);
      if (legacy != null && legacy.isNotEmpty) {
        await _db.setKv(kvKey, legacy);
        await _secure.delete(key: _legacySecureKey);
        return _cacheFromBase64(legacy);
      }
    } catch (e) {
      debugPrint('Field crypto secure-storage migrate failed: $e');
    }

    // Dev fallback only when no compile-time key.
    final generated = FieldCrypto.generateKeyBase64();
    await _db.setKv(kvKey, generated);
    debugPrint(
      'Field crypto: generated local key (set FIELD_CRYPTO_KEY_BASE64 '
      'for shared key across installs/.NET)',
    );
    return _cacheFromBase64(generated);
  }

  FieldCrypto? _cacheFromBase64(String keyBase64) {
    try {
      _cached = FieldCrypto.fromBase64Key(keyBase64);
      return _cached;
    } catch (e) {
      debugPrint('Invalid field crypto key: $e');
      return null;
    }
  }

  Future<void> clearLocal() async {
    _cached = null;
    await _db.deleteKv(kvKey);
    try {
      await _secure.delete(key: _legacySecureKey);
    } catch (_) {}
  }
}

File: app/src/main/java/com/example/salik_management_system/core/crypto/FieldCryptoKeyStore.kt
package com.example.salik_management_system.core.crypto

import android.util.Log
import com.example.salik_management_system.auth.domain.UserSession
import com.example.salik_management_system.core.config.AppConfig
import com.example.salik_management_system.core.database.LocalAppKvDao
import com.example.salik_management_system.core.database.LocalAppKvEntity
import javax.inject.Inject
import javax.inject.Singleton

/**
 * AES field-crypto key stored in Room `local_app_kv` only (never Firestore).
 * Prefers [AppConfig.fieldCryptoKeyBase64] when set (shared org key).
 */
@Singleton
class FieldCryptoKeyStore @Inject constructor(
    private val kvDao: LocalAppKvDao,
) {
    @Volatile
    private var cached: FieldCrypto? = null

    val current: FieldCrypto? get() = cached

    suspend fun ensureKey(session: UserSession? = null): FieldCrypto? {
        cached?.let { return it }

        val fromConfig = AppConfig.fieldCryptoKeyBase64.trim()
        if (fromConfig.isNotEmpty()) {
            val existing = kvDao.get(KV_KEY)
            if (existing != fromConfig) {
                kvDao.upsert(LocalAppKvEntity(KV_KEY, fromConfig))
            }
            return cacheFromBase64(fromConfig)
        }

        val fromKv = kvDao.get(KV_KEY)
        if (!fromKv.isNullOrEmpty()) {
            return cacheFromBase64(fromKv)
        }

        val generated = FieldCrypto.generateKeyBase64()
        kvDao.upsert(LocalAppKvEntity(KV_KEY, generated))
        Log.d(TAG, "Field crypto: generated local key")
        return cacheFromBase64(generated)
    }

    private fun cacheFromBase64(keyBase64: String): FieldCrypto? {
        return try {
            FieldCrypto.fromBase64Key(keyBase64).also { cached = it }
        } catch (e: Exception) {
            Log.d(TAG, "Invalid field crypto key: $e")
            null
        }
    }

    companion object {
        const val KV_KEY = "field_crypto_key_v1"
        private const val TAG = "FieldCryptoKeyStore"
    }
}
